File: homework3/ex1.py
import random

rates = [1, 1/2, 1/3, 1/4, 1/5, 1/8]
p = 1/10

def random_N_R_code_distinct(n, r):
    """
    Create a random binary code with codeword length N and rate R, with
    distinct codewords.
    """
    num_of_codewords = 2**int(n*r)

    # random.sample() fails for possible codewords = 2**64 with OverflowError: Python int too large to convert to C ssize_t
    randomly_picked_codewords = set()
    while len(randomly_picked_codewords) < num_of_codewords:
        randomly_picked_codewords.add(random.randint(0, 2**n))

    return [[1 if bit=='1' else 0 for bit in f"{codeword:0{n}b}"] for codeword in randomly_picked_codewords]


def random_N_R_code(n, r):
    """
    Create a random binary code with codeword length N and rate R.
    """
    # we code NR bits
    num_of_codewords = 2**int(n*r)

    # we create 2**NR codewords at random
    codewords = [[0 for _ in range(n)] for _ in range(num_of_codewords)]
    for i in range(num_of_codewords):
        for j in range(n):
            x_i = random.randint(0, 1)
            codewords[i][j] = x_i

    return codewords

def channel(x, p):
    output = [0] * len(x)
    for i, b in enumerate(x):
        output[i] = b if random.random() >= p else int(not b)
    return output

def ML_decode(received, inputs):
    min_dist = sum([1 for x, y in zip(received, inputs[0]) if x!=y])
    estimate = inputs[0]
    for input in inputs[1:]:
        hamming_dist = sum([1 for x, y in zip(received, input) if x!=y])
        if hamming_dist < min_dist:     # assuming no equalities
            min_dist = hamming_dist
            estimate = input
    return estimate

for rate in rates:

    unit = int(1/rate)

    bit_errors = [0 for _ in range(10)]

    for n in range(unit, 10*unit+1, unit):
        codewords = random_N_R_code_distinct(n, rate)
        print(f"rate {rate}, n {n}, unit {unit}, {len(codewords)} ({2**int(n*rate)}) codewords, {len(set([''.join(map(str, word)) for word in codewords]))} distinct")

        for i in range(100_000):
            print(f"experiment {i}", end="\r", flush=True)

            cw_idx = random.randint(0, len(codewords)-1)
            cw = codewords[cw_idx]

            channeled = channel(cw, p)
            decoded = ML_decode(channeled, codewords)
            _bit_errors = sum([1 for x, y in zip(cw, decoded) if x!=y])

            bit_errors[int(n/unit)-1] += _bit_errors
    print(f"bit errors for rate {rate}: {bit_errors}")

File: homework3/ex_1.m
rates = [1, 1/2, 1/3, 1/4, 1/5, 1/8];
p = 1/10;
num_of_experiments = 100000;
bit_error_rates = zeros(length(rates), 10);

for rate_idx = 1:length(rates)
    rate = rates(rate_idx);
    unit = floor(1 / rate);

    for n = unit:unit:10*unit
        k = floor(n*rate);
        disp(['rate ', num2str(rate), ', n ', num2str(n), ', unit ', num2str(unit), ...
              ', (', num2str(2^k), ') codewords']);

        n_r_bit_errors = 0;
        parfor i = 1:num_of_experiments            % can be replaced with parfor for faster execution
            random_code = random_N_R_code(n, rate);
            experiment_bit_errors = 0;
            for j = 1:10

                message = randi([0 1], k, 1);                           % generate random k-bit message
                cw_idx = bit2int(message, k)+1;                         % make it decimal so it can be used as an index
                cw = random_code(cw_idx, :);                            % get the codeword of that index (i.e. of that message)
                channeled = channel(cw, p);                             % transmit it
                msg_estimate = ML_decode(channeled, random_code, k);    % estimate the message based on the noisy codeword with ML 

                bit_errors = sum(message ~= msg_estimate);
                experiment_bit_errors = experiment_bit_errors + bit_errors;
            end
            n_r_bit_errors = n_r_bit_errors + experiment_bit_errors;
        end
        bit_error_rates(rate_idx, n/unit) = bit_error_rates(rate_idx, n/unit) + n_r_bit_errors/(n*num_of_experiments*10);
    end
    disp(['bit error rates for rate ', num2str(rate), ': ',  mat2str(bit_error_rates(rate_idx, :))]);
end


function codewords = random_N_R_code_distinct(n, r)
    num_codewords = 2^floor(n*r);
    codeword_set = zeros(num_codewords, n);
    picked = containers.Map('KeyType', 'char', 'ValueType', 'logical');
    i = 1;
    while i <= num_codewords
        bits = randi([0, 1], 1, n);
        bin_str = char(bits + '0');
        key = string(bin_str);
        if ~isKey(picked, key)
            picked(key) = true;
            codeword_set(i, :) = bits;
            i = i + 1;
        end
    end
    codewords = codeword_set;
end

function codewords = random_N_R_code(n, r)
        num_codewords = 2^floor(n*r);
        codewords = randi([0,1], num_codewords, n);
end

function output = channel(x, p)
    flip = rand(1, length(x)) < p;
    output = xor(x, flip);
end

function estimate = ML_decode(received, inputs, k)
    distances = sum(received ~= inputs, 2);
    [~, min_idx] = min(distances);
    estimate = int2bit(min_idx+1, k); % make the estimated codeword index back
end


figure;
hold on;

for i = 1:length(rates)
    plot(1:10, bit_error_rates(i, :), DisplayName=['R = ' num2str(rates(i), 3)], LineWidth=1.5);
end

title('Evolution of Bit Error Rate with Codeword Length');
xlabel('Sequence Length (in bits, xR^{-1})');
ylabel('Bit Error Rate');
legend('show');
grid on;
hold off;

